src/hooks/useQuestionnaireResponseDraft.ts

```diff
@@ -0,0 +1,281 @@
+    deleteDraft: () => Promise<void>;
+    draftQuestionnaireResponseRD: RemoteData<WithId<QuestionnaireResponse> | undefined>;
+    draftInfoMessage?: string;
+    onUpdateDraft: (questionnaireResponse: QuestionnaireResponse) => Promise<void>;
+}
+
+export const useQuestionnaireResponseDraft = (
+    props: QuestionnaireResponseDraftProps,
+): QuestionnaireResponseDraftResponse => {
+    const {
+        subject,
+        questionnaireId,
+        questionnaireResponseId,
+        qrDraftServiceType = 'local',
+        autoSave = false,
+        questionnaireResponse,
+    } = props;
+
+    const [draftInfoMessage, setDraftInfoMessage] = useState<string | undefined>();
+    const draftKeyRef = useRef<string | undefined>();
+
+    const draftKeyPrefix = useMemo(
+        () =>
+            getQuestionnaireResponseDraftKeyPrefix({
+                subject,
+                questionnaireResponseId,
+                qrDraftServiceType,
+            }),
+        [subject, questionnaireResponseId, qrDraftServiceType],
+    );
+
+    const draftId = useMemo(() => questionnaireResponse?.id ?? uuid4(), [questionnaireResponse]);
+
+    const [response, manager] = useService<WithId<QuestionnaireResponse>>(async () => {
+        const questionnaireRD = await getFHIRResources<Questionnaire>('Questionnaire', {
+            id: questionnaireId,
+            _elements: ['id', 'meta'].join(','),
+        });
+
+        if (isFailure(questionnaireRD)) {
+            return failure(t`Questionnaire not found`);
+        }
+
+        const questionnaire = extractBundleResources(questionnaireRD.data).Questionnaire[0];
+
+        const questionnaireURI = `Questionnaire/${questionnaire?.id}/_history/${questionnaire?.meta?.versionId}`;
+
+        draftKeyRef.current = `${draftKeyPrefix}|${questionnaireURI}`;
+
+        return mapSuccess(
+            await loadQuestionnaireResponseDraft(draftKeyRef.current, qrDraftServiceType),
+
+            (draftQR) => {
+                setDraftInfoMessage(
+                    t`Draft was successfully loaded from ${
+                        qrDraftServiceType === 'local' ? 'local storage' : 'FHIR server'
+                    }`,
+                );
+                return draftQR;
+            },
+        );
+    }, [draftKeyPrefix, qrDraftServiceType]);
+
+    const saveDraft = useCallback(
+        async (questionnaireResponse: QuestionnaireResponse): Promise<RemoteDataResult<QuestionnaireResponse>> => {
+            if (!autoSave || !questionnaireId || !questionnaireResponse) {
+                const reason = !autoSave
+                    ? t`Auto save is disabled`
+                    : !questionnaireId
+                    ? t`Questionnaire id is required`
+                    : t`Form data is required`;
+                return failure(reason);
+            }
+
+            const draftQRRD = await saveQuestionnaireResponseDraft(
+                draftKeyRef.current,
+                draftId,
```

Comment:
don't pass draftId, use questionnaireResponse.id instead that 100% is defined (because of useService)